### ISSUE TITLE
Feature Request: Local response normalization (LRN)

### ISSUE THREAD

**Issue by keunhong** [pytorch/pytorch#653]
PyTorch doesn't seem to have a module for LRN yet. This is needed for running models pre-trained with LRN layers (mainly AlexNet based models). I have a workaround using the legacy API

```python
from torch import nn
from torch.autograd import Variable
from torch.legacy.nn import SpatialCrossMapLRN


class LRN(nn.Module):
    def __init__(self, size, alpha=1e-4, beta=0.75, k=1):
        super().__init__()
        self.lrn = SpatialCrossMapLRN(size, alpha, beta, k)

    def forward(self, x):
        self.lrn.clearState()
        return Variable(self.lrn.updateOutput(x.data))
```

but that doesn't seem to have a CUDA binding.

**Comment by apaszke** [comment]
Just keep in mind that it won't support backpropagation to `x` if you implemented it this way. We'll add LRN soon.

**Comment by thnkim** [comment]
I'm waiting for LRN also.

**Comment by Eniac-Xie** [comment]
waiting for LRN, too

**Comment by lihao1992** [comment]
I just want to running some model based on GoogLeNet, but i haven't found LRN module in Pytorch. I wish anyone can fix it or add it , please

**Comment by playerkk** [comment]
Another request of the LRN layer. Thanks.

**Comment by fmassa** [comment]
Here is a workaround implementation:
```python
import torch
from torch.legacy.nn import SpatialCrossMapLRN as SpatialCrossMapLRNOld
from torch.autograd import Function, Variable
from torch.nn import Module

# function interface, internal, do not use this one!!!
class SpatialCrossMapLRNFunc(Function):
    def __init__(self, size, alpha=1e-4, beta=0.75, k=1):
        self.size = size
        self.alpha = alpha
        self.beta = beta
        self.k = k

    def forward(self, input):
        self.save_for_backward(input)
        self.lrn = SpatialCrossMapLRNOld(self.size, self.alpha, self.beta, self.k)
        self.lrn.type(input.type())
        return self.lrn.forward(input)

    def backward(self, grad_output):
        input, = self.saved_tensors
        return self.lrn.backward(input, grad_output)

# use this one instead
class SpatialCrossMapLRN(Module):
    def __init__(self, size, alpha=1e-4, beta=0.75, k=1):
        super(SpatialCrossMapLRN, self).__init__()
        self.size = size
        self.alpha = alpha
        self.beta = beta
        self.k = k

    def forward(self, input):
        return SpatialCrossMapLRNFunc(self.size, self.alpha, self.beta, self.k)(input)
```

It uses the legacy implementation, and might not be great because it might keep some tensors around. But seems to work and passes grad-checks (but was not extensively tested). Use with care :)

**Comment by zym1010** [comment]
@fmassa So I have tried your code, and found that, on 0.12, cpu and cuda give different results. Do you notice that?

**Comment by fmassa** [comment]
@zym1010 I have not used the python version of LRN, but I wrote the lua one, and if I recall properly it was consistent with the cuda implementation. How different are the results?

**Comment by zym1010** [comment]
@fmassa check #1938. It's much less stable, compared to Caffe's CPU implementation

**Comment by vadimkantorov** [comment]
@fmassa I've been using [CrossMapLRN2d]( https://github.com/pytorch/pytorch/blob/master/torch/nn/_functions/thnn/normalization.py#L8), though Soumith mentioned that it [shouldn't be used](https://discuss.pytorch.org/t/how-to-use-crossmaplrn2d/2894). Do you know the status of this CrossMapLRN2d code?

**Comment by shanyuhu** [comment]
Has the LRN layer been added into pytorch now? really need it.

**Comment by jiecaoyu** [comment]
A simple implementation for LRN in AlexNet (only ACROSS_CHANNELS case):

```python
class LRN(nn.Module):
    def __init__(self, local_size=1, alpha=1.0, beta=0.75, ACROSS_CHANNELS=True):
        super(LRN, self).__init__()
        self.ACROSS_CHANNELS = ACROSS_CHANNELS
        if ACROSS_CHANNELS:
            self.average=nn.AvgPool3d(kernel_size=(local_size, 1, 1), 
                    stride=1,
                    padding=(int((local_size-1.0)/2), 0, 0)) 
        else:
            raise Exception('Only ACROSS_CHANNELS is supported.')
        self.alpha = alpha
        self.beta = beta


    def forward(self, x): 
        if self.ACROSS_CHANNELS:
            div = x.pow(2).unsqueeze(1)
            div = self.average(div).squeeze(1)
            div = div.mul(self.alpha).add(1.0).pow(self.beta)
        else:
            raise Exception('Only ACROSS_CHANNELS is supported.')
        x = x.div(div)
        return x
```

**Comment by eriche2016** [comment]
When will LRN  layer be available?  By the way, @jiecaoyu  has the code being tested?

**Comment by jiecaoyu** [comment]
@eriche2016 I have used the LRN in an AlexNet implementation ([link](https://github.com/jiecaoyu/pytorch_imagenet/blob/master/networks/model_list/alexnet.py)). Check [here](https://github.com/jiecaoyu/pytorch_imagenet) for more details.

**Comment by eriche2016** [comment]
@jiecaoyu  Thank you very much for your work.

**Comment by hdmetor** [comment]
I can take a look at this too, if there is still interest

**Comment by soumith** [comment]
@hdmetor that'd be nice.

**Comment by hdmetor** [comment]
Tracking here

https://github.com/hdmetor/pytorch/tree/LRN

**Comment by hdmetor** [comment]
created https://github.com/pytorch/pytorch/pull/4237

**Comment by TaihuLight** [comment]
@apaszke @hdmetor @jiecaoyu 
How to implement a stable LRN layer in Pytorch 0.3 with Python 3.5? 
BTW, I hope the LRN layer is add in the next version (0.4) of the official Pytorch!

**Comment by soumith** [comment]
closed via https://github.com/pytorch/pytorch/pull/4667
